src/java.base/share/classes/java/lang/reflect/Executable.java

```diff
@@ -421,46 +421,40 @@ private void verifyParameters(final Parameter[] parameters) {
         }
     }
 
-    private Parameter[] privateGetParameters() {
-        // Use tmp to avoid multiple writes to a volatile.
-        Parameter[] tmp = parameters;
 
-        if (tmp == null) {
-
-            // Otherwise, go to the JVM to get them
-            try {
-                tmp = getParameters0();
-            } catch(IllegalArgumentException e) {
-                // Rethrow ClassFormatErrors
-                throw new MalformedParametersException("Invalid constant pool index");
-            }
-
-            // If we get back nothing, then synthesize parameters
-            if (tmp == null) {
-                hasRealParameterData = false;
-                tmp = synthesizeAllParams();
-            } else {
-                hasRealParameterData = true;
-                verifyParameters(tmp);
-            }
+    boolean hasRealParameterData() {
+        return parameterData().isReal;
+    }
 
-            parameters = tmp;
+    private ParameterData parameterData() {
+        ParameterData parameterData = this.parameterData;
+        if (parameterData != null) {
+            return parameterData;
         }
 
-        return tmp;
-    }
+        Parameter[] tmp;
+        // Go to the JVM to get them
+        try {
+            tmp = getParameters0();
+        } catch (IllegalArgumentException e) {
+            // Rethrow ClassFormatErrors
+            throw new MalformedParametersException("Invalid constant pool index");
+        }
 
-    boolean hasRealParameterData() {
-        // If this somehow gets called before parameters gets
-        // initialized, force it into existence.
-        if (parameters == null) {
-            privateGetParameters();
+        // If we get back nothing, then synthesize parameters
+        if (tmp == null) {
+            tmp = synthesizeAllParams();
+            parameterData = new ParameterData(tmp, false);
+        } else {
+            verifyParameters(tmp);
+            parameterData = new ParameterData(tmp, true);
         }
-        return hasRealParameterData;
+        return this.parameterData = parameterData;
     }
 
-    private transient volatile boolean hasRealParameterData;
-    private transient volatile Parameter[] parameters;
+    private transient @Stable ParameterData parameterData;
+
+    record ParameterData(@Stable Parameter[] parameters, boolean isReal) {}
```

Comment:
Record fields are implicitly "trusted final", which means they don't need @Stable annotation:
https://github.com/openjdk/jdk/blob/124c63c17c897404e3c5c3615d6727303e4f3d06/src/hotspot/share/ci/ciField.cpp#L240

Reply:
The [`@Stable`] annotation is there to make the array entries as being stable:
https://github.com/openjdk/jdk/blob/d260a4e794681c6f4be4767350702754cfc2035c/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L51-L53 https://github.com/openjdk/jdk/blob/d260a4e794681c6f4be4767350702754cfc2035c/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L58-L62

[`@Stable`]: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

Reply:
@plevart so should I remove it or keep as is?

Reply:
Right, in that case, it should remain.

Reply:
But, ... is any code path accessing the elements of the @Stable array by constant indexes? Only in that case would the annotation have any effect on the JIT-ed code. Otherwise it's just a waste of space.

Reply:
... I can only see the array being cloned and not accessed directly. I don't belive cloning a @stable array is any different in JIT-ed code as cloning normal "mutable" array unless JIT "sees" through it and scalarizes the values of the array. For example, if you have the following:

```
static final Method method = ....;

@Benchmark
public Object getParameter0() {
  return method.getParameters()[0];
}
```

...would it run faster when the parameters field was marked as @stable as opposed to not?

Reply:
@plevart I've checked it with and without `@stable`, it's the same:
```
with
Benchmark                            Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.getParameter0  avgt   50  6,196 ± 0,142  ns/op
AccessParamsBenchmark.getParameters  avgt   50  4,636 ± 0,075  ns/op

without
Benchmark                            Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.getParameter0  avgt   50  6,196 ± 0,142  ns/op
AccessParamsBenchmark.getParameters  avgt   50  4,636 ± 0,075  ns/op
```
This seems logical as effects of `@Stable` aren not propagated into cloned array.
This is the benchmark I used:
```java
@State(Scope.Thread)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Fork(jvmArgsAppend = {"-Xms1g", "-Xmx1g"})
public class AccessParamsBenchmark {

  private final Method method;

  public AccessParamsBenchmark() {
    try {
      method = getClass().getMethod("foo", int.class, String.class);
    } catch (NoSuchMethodException e) {
      throw new RuntimeException(e);
    }
  }

  @Benchmark
  public Object getParameters() {
    return method.getParameters();
  }

  @Benchmark
  public Object getParameter0() {
    return method.getParameters()[0];
  }

  public void foo(int parameter1, String parameter2) {
  }
}
```

So, should we rid the annotation from `parameters` field?

Reply:
@Stable is only effective if the path leading to @Stable value can be constant-folded by JIT. In above test, you have an instance field Method method. This can not be constant-folded, so neither can @stable fiels in the Field object, nor array elements of a @stable array. You should replace that with "static final Method method = ..." and re-run the test.

Reply:
With `static` and without `@Stable` the benchmark yields
```
Benchmark                            Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.getParameter0  avgt   10  1,212 ± 0,083  ns/op
AccessParamsBenchmark.getParameters  avgt   10  2,493 ± 0,076  ns/op
```
and with `@Stable`
```
Benchmark                            Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.getParameter0  avgt   40  0,427 ± 0,007  ns/op
AccessParamsBenchmark.getParameters  avgt   40  2,123 ± 0,052  ns/op
```
so the annotation is useful.

Reply:
hmm, is the faster getParameters (without explicit index access) a result of the annotation? getParameter0 shows the documented effect but isn't quite our use case here.

Reply:
...neither is obtaining a cloned array and passing its reference to JMH's black hole our usecase... Still, it seems that even part of that (obtaining the source array reference as a constant vs. variable before cloning the referent has some advantage). I would keep the @stable annotation then.

Reply:
A more realistic use case would be something in the lines of checking whether the method is an expected one:

```java
  @Benchmark
  public boolean isFoo() {
    return matches(method, "foo", int.class, String.class);
  }
  
  private boolean matches(Method method, String name, Class<?> ...parameterTypes) {
    if (method.getName().equals(name) &&
        method.getParameterCount() == parameterTypes.length) {
      var params = method.getParameters();
      for (int i = 0; i < parameterTypes.length; i++) {
        if (params[i].getType() != parameterTypes[i]) {
          return false;
        }
      }
      return true;
    } else {
      return false;
    }
  }
```

Reply:
Without `@Stable` your benchmark yields
```
Benchmark                    Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.isFoo  avgt   40  1,110 ± 0,062  ns/op
```
and with `@Stable` it yields
```
Benchmark                    Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.isFoo  avgt   40  0,836 ± 0,015  ns/op
```
Java 18:
```
Benchmark                    Mode  Cnt  Score   Error  Units
AccessParamsBenchmark.isFoo  avgt   40  6,105 ± 0,107  ns/op
```

Reply:
So, it looks like @Stable should stay.